src/ctapipe/monitoring/calculator.py

```diff
@@ -0,0 +1,336 @@
+"""
+Definition of the ``PixelStatisticsCalculator`` class, providing all steps needed to
+calculate the montoring data for the camera calibration.
+"""
+
+import numpy as np
+from astropy.table import Table, vstack
+
+from ctapipe.core import TelescopeComponent
+from ctapipe.core.traits import (
+    ComponentName,
+    Dict,
+    Float,
+    Int,
+    List,
+    TelescopeParameter,
+    TraitError,
+)
+from ctapipe.monitoring.aggregator import StatisticsAggregator
+from ctapipe.monitoring.outlier import OutlierDetector
+
+__all__ = [
+    "PixelStatisticsCalculator",
+]
+
+
+class PixelStatisticsCalculator(TelescopeComponent):
+    """
+    Component to calculate statistics from calibration events.
+
+    The ``PixelStatisticsCalculator`` is responsible for calculating various statistics from
+    calibration events, such as pedestal and flat-field data. It aggregates statistics,
+    detects outliers, and handles faulty data periods.
+    This class holds two functions to conduct two different passes over the data with and without
+    overlapping aggregation chunks. The first pass is conducted with non-overlapping chunks,
+    while overlapping chunks can be set by the ``chunk_shift`` parameter for the second pass.
+    The second pass over the data is only conducted in regions of trouble with a high fraction
+    of faulty pixels exceeding the threshold ``faulty_pixels_threshold``.
+    """
+
+    stats_aggregator_type = TelescopeParameter(
+        trait=ComponentName(
+            StatisticsAggregator, default_value="SigmaClippingAggregator"
+        ),
+        default_value="SigmaClippingAggregator",
+        help="Name of the StatisticsAggregator subclass to be used.",
+    ).tag(config=True)
+
+    outlier_detector_list = List(
+        trait=Dict(),
+        default_value=None,
+        allow_none=True,
+        help=(
+            "List of dicts containing the name of the OutlierDetector subclass to be used, "
+            "the aggregated statistic value to which the detector should be applied, "
+            "and the validity range of the detector. "
+            "E.g. ``[{'apply_to': 'std', 'name': 'RangeOutlierDetector', 'validity_range': [2.0, 8.0]},]``."
+        ),
+    ).tag(config=True)
+
+    chunk_shift = Int(
+        default_value=None,
+        allow_none=True,
+        help=(
+            "Number of samples to shift the aggregation chunk for the calculation "
+            "of the statistical values. Only used in the second_pass(), since the "
+            "first_pass() is conducted with non-overlapping chunks (chunk_shift=None)."
+        ),
+    ).tag(config=True)
+
+    faulty_pixels_threshold = Float(
+        default_value=0.1,
+        allow_none=True,
+        help="Threshold of faulty pixels over the camera to identify regions of trouble.",
```

Comment:
I don't quite understand from the help what the units are here. Is this a fraction of the camera? If so, maybe rename it to `faulty_pixels_fraction` and say "Minimum fraction of faulty camera pixels...."

Reply:
Sounds good! Resolved in the latest commit.